### PR TITLE
Information disclosure tainted data DFA

### PR DESCRIPTION
**Pull request by dotpaul**
- Now SinkInfos can be for multiple SinkKinds
- Hard coded sources and sinks (from FxCop, minus the enum value sinks) for information disclosure
- Smoke tests for tainted data DFA information disclosure

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/Core/Security/MicrosoftNetCoreSecurityResources.resx [hunk 1]
```diff
@@ -171,4 +171,10 @@
   <data name="ReviewCodeForDllInjectionVulnerabilitiesTitle" xml:space="preserve">
     <value>Review code for DLL injection vulnerabilities</value>
   </data>
+  <data name="ReviewCodeForInformationDisclosureVulnerabilitiesMessage" xml:space="preserve">
+    <value>Potential information disclosure vulnerability was found where '{0}' in method '{1}' may be contain unintended information from '{2}' in method '{3}'.</value>
```
Comment (timhermann) [inline src/Microsoft.NetCore.Analyzers/Core/Security/MicrosoftNetCoreSecurityResources.resx]: "may be contain" => "may contain"

File: src/Utilities/Extensions/ImmutableHashSetBuilderExtensions.cs [hunk 2]
```diff
@@ -10,31 +11,70 @@ namespace Analyzer.Utilities.Extensions
...
+            IEnumerable<(string Method, string[] Parameters)> sinkMethodParameters)
         {
             SinkInfo sinkInfo = new SinkInfo(
                 fullTypeName,
-                sinkKind,
+                sinkKinds.ToImmutableHashSet(),
                 isInterface,
                 isAnyStringParameterInConstructorASink,
                 sinkProperties:
                     sinkProperties != null
```
Comment (timhermann) [inline src/Utilities/Extensions/ImmutableHashSetBuilderExtensions.cs]: suggestion: `sinkProperties?.ToImmutableHashSet(StringComparer.Ordinal) ?? ImmutableHashSet<string>.Empty`

File: src/Utilities/Extensions/ImmutableHashSetBuilderExtensions.cs [hunk 3]
```diff
@@ -10,31 +11,70 @@ namespace Analyzer.Utilities.Extensions
...
+            this ImmutableHashSet<SourceInfo>.Builder builder,
+            string fullTypeName,
+            bool isInterface,
+            string[] taintedProperties,
+            string[] taintedMethods)
+        {
+            SourceInfo metadata = new SourceInfo(
+                fullTypeName,
+                isInterface: isInterface,
+                taintedProperties: taintedProperties != null
```
Comment (timhermann) [inline src/Utilities/Extensions/ImmutableHashSetBuilderExtensions.cs]: same suggestion as above - null coalesce properties and supply a default for when the null shortcut is taken

File: src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs [hunk 4]
```diff
@@ -435,7 +435,7 @@ private bool IsMethodArgumentASink(IMethodSymbol method, IEnumerable<IArgumentOp
                 Lazy<HashSet<SinkKind>> lazySinkKinds = new Lazy<HashSet<SinkKind>>(() => new HashSet<SinkKind>());
                 foreach (SinkInfo sinkInfo in this.DataFlowAnalysisContext.SinkInfos.GetInfosForType(method.ContainingType))
```
Comment (timhermann) [inline src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs]: suggestion: `var sinkInfos = this.DataFlowAnalysisContext.SinkInfos.GetInfosForType(method.ContainingType).Where(si => !(lazySinkKinds.IsValueCreated && lazySinkKinds.Value.IsSupersetOf(si));`  then `foreach (var sinkInfo in sinkInfos)`

Comment (dotpaul) [reply src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs]: I think I'd find that hard to follow, and `lazySinkKinds` is mutable while iterating, so I don't think it belongs in `Where()`.

**Pull request merged**